src/client/datascience/data-viewing/dataViewer.ts

```diff
@@ -99,23 +114,31 @@ export class DataViewer extends WebviewPanelHost<IDataViewerMapping> implements
     protected onMessage(message: string, payload: any) {
         switch (message) {
             case DataViewerMessages.GetAllRowsRequest:
-                this.getAllRows().ignoreErrors();
+                this.getAllRows(payload as string).ignoreErrors();
                 break;
 
             case DataViewerMessages.GetRowsRequest:
                 this.getRowChunk(payload as IGetRowsRequest).ignoreErrors();
                 break;
 
+            case DataViewerMessages.GetSliceRequest:
+                this.getSlice(payload as IGetSliceRequest).ignoreErrors();
+                break;
+
             default:
                 break;
         }
 
         super.onMessage(message, payload);
     }
 
-    private getDataFrameInfo(): Promise<IDataFrameInfo> {
-        if (!this.dataFrameInfoPromise) {
-            this.dataFrameInfoPromise = this.dataProvider ? this.dataProvider.getDataFrameInfo() : Promise.resolve({});
+    private getDataFrameInfo(sliceExpression?: string): Promise<IDataFrameInfo> {
+        // If requesting a new slice, refresh our cached info promise
+        if (!this.dataFrameInfoPromise || sliceExpression !== this.currentSlice) {
```

Comment:
This implies that previous slicing has not yet completed and the user chose a new slice expression.
Does this mean we need to support the concept of cancellations?
Is it possible that slicing can take a while (in which case a cancellation might prove useful, & possibly complicate the entire messaging as well)?

Reply:
Not necessarily, this is here because we used to cache `dataFrameInfoPromise` (to avoid running our Python code every time we needed to get the info). We need to force a refetch when the slice being requested isn't already cached.

It is possible that slicing can take a while. One solution for that would be disabling the UI when a slice is in progress. But this PR doesn't handle UI, as that hasn't been designed yet.

Reply:
Cancellations would be nice but just making slicing a blocking procedure from the UI would work for a start. I think that I'm following now. I guess I would argue that (until we support cancellations) this might be more of a traceError case? The state where there is an outstanding dataFrameInfoPromise but the slice expression has changed seems like an invalid state to me. It would be unclear if getDataFrameInfo would be returning the results from the previous dataFrameInfoPromise or from the current slice expression.

Reply:
I'm afraid I'm not quite following why this is invalid. If there is a cached promise and the slice expression has changed, that just means the user has asked for a new slice, and we should refetch, not reuse the cached promise.

Reply:
I agree with Ian, 
Its possible i add a slice expression & then immediately change it to something else.
Its possibel I'd get the first result (even though the cached promise was cleared) & then later I'd get the next restul.

This is generally handled at the UI that makes the request (i.e. ignore the last request sent, kind of how debouncing works in UI when making network calls, throw away previous requests), where we cancel the output from the previous results.

One way of addressing this is to pass a request Id to the extension side.
Then instead of returning the promise, we await on the results in the function. & if the latest request id is the same as the request id for which we have a response, then we send that, else await on the latest promise in the cached property.

Note: But this all feels overly complicated, again, this is generally handled at the UI layer (Redux, ReduxThunk or vanilla JS with debounce solves this)..

I personally don't think we need to implement anything to address this issue.
I.e. I'm happy with what you have today (if we did want to fix it, i'd fix it at the UI end & introduce cancellations to cancel previous requests in backend to save CPU - but not required now, just complicate it).

Reply:
Once again, I don't think we need to fix anything here.

Reply:
I understand the scenario now, thanks for explaining. Whichever UI we settle on, I'll make sure I address pending slice requests in part 2. The request ID approach feels pretty intuitive to me.